File: examplepatches/roflsunriz/google-app-revanced/extensions/googleapp/src/main/java/app/revanced/extension/googleapp/ResourceClassifier.java
package app.revanced.extension.googleapp;

import java.util.Locale;
import java.util.regex.Pattern;

public final class ResourceClassifier {
    private static final Pattern AD_TOKEN = Pattern.compile(
            "(^|_)(ad|ads|advert|advertisement|sponsor|sponsored)(_|$)"
    );
    private static final Pattern PROMO_TOKEN = Pattern.compile(
            "(^|_)(promo|promoted|promotion|promotional)(_|$)"
    );

    private ResourceClassifier() {
    }

    public static boolean isAdName(String value) {
        String normalized = normalize(value);
        return AD_TOKEN.matcher(normalized).find()
                || normalized.contains("admob")
                || normalized.contains("doubleclick")
                || normalized.contains("google_ads");
    }

    public static boolean isPromotionName(String value) {
        return PROMO_TOKEN.matcher(normalize(value)).find();
    }

    private static String normalize(String value) {
        return value == null ? "" : value.toLowerCase(Locale.ROOT).replace('-', '_');
    }
}

File: examplepatches/roflsunriz/google-app-revanced/patches/src/main/kotlin/dev/roflsunriz/googleapp/ResourceClassifier.kt
package dev.roflsunriz.googleapp

internal object ResourceClassifier {
    private val adToken = Regex("(^|_)(ad|ads|advert|advertisement|sponsor|sponsored)(_|$)")
    private val promoToken = Regex("(^|_)(promo|promoted|promotion|promotional)(_|$)")

    fun isAdName(name: String): Boolean {
        val normalized = normalize(name)
        return adToken.containsMatchIn(normalized) ||
            normalized.contains("admob") ||
            normalized.contains("doubleclick") ||
            normalized.contains("google_ads")
    }

    fun isPromotionName(name: String): Boolean = promoToken.containsMatchIn(normalize(name))

    fun shouldCollapseLayout(name: String): Boolean = isAdName(name) || isPromotionName(name)

    private fun normalize(name: String) = name.lowercase().replace('-', '_')
}
